File: go-api/internal/controller/middleware/jwt_test.go
package middleware

import (
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v5"
)

const testSecret = "test-secret"

func signToken(t *testing.T, secret string, claims jwt.RegisteredClaims) string {
	t.Helper()
	token, err := jwt.NewWithClaims(jwt.SigningMethodHS256, claims).SignedString([]byte(secret))
	if err != nil {
		t.Fatalf("sign token: %v", err)
	}
	return token
}

func newTestApp() *fiber.App {
	app := fiber.New()
	app.Use(JWTProtected(testSecret))
	app.Get("/protected", func(c *fiber.Ctx) error {
		return c.SendString("ok")
	})
	return app
}

func TestJWTProtected(t *testing.T) {
	validClaims := jwt.RegisteredClaims{
		Subject:   "test-user",
		IssuedAt:  jwt.NewNumericDate(time.Now()),
		ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Hour)),
	}

	t.Run("valid token is let through", func(t *testing.T) {
		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Bearer "+signToken(t, testSecret, validClaims))

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusOK {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusOK)
		}
	})

	t.Run("missing Authorization header is rejected", func(t *testing.T) {
		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})

	t.Run("non-Bearer scheme is rejected", func(t *testing.T) {
		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Basic dXNlcjpwYXNz")

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})

	t.Run("garbage token is rejected", func(t *testing.T) {
		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Bearer not.a.jwt")

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})

	t.Run("token signed with the wrong secret is rejected", func(t *testing.T) {
		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Bearer "+signToken(t, "wrong-secret", validClaims))

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})

	t.Run("expired token is rejected", func(t *testing.T) {
		expired := jwt.RegisteredClaims{
			Subject:   "test-user",
			IssuedAt:  jwt.NewNumericDate(time.Now().Add(-2 * time.Hour)),
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(-time.Hour)),
		}

		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Bearer "+signToken(t, testSecret, expired))

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})

	t.Run("alg:none token is rejected (algorithm confusion)", func(t *testing.T) {
		// alg:none should never be accepted
		unsigned, err := jwt.NewWithClaims(jwt.SigningMethodNone, validClaims).SignedString(jwt.UnsafeAllowNoneSignatureType)
		if err != nil {
			t.Fatalf("build alg:none token: %v", err)
		}

		app := newTestApp()
		req := httptest.NewRequest(http.MethodGet, "/protected", nil)
		req.Header.Set("Authorization", "Bearer "+unsigned)

		resp, err := app.Test(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if resp.StatusCode != http.StatusUnauthorized {
			t.Errorf("status = %d, want %d", resp.StatusCode, http.StatusUnauthorized)
		}
	})
}

File: node-api/src/controller/middleware/jwt.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import express from 'express'
import jwt from 'jsonwebtoken'
import { jwtProtected } from './jwt.js'

const SECRET = 'test-secret'

function startTestApp() {
  const app = express()
  app.use(jwtProtected(SECRET))
  app.get('/protected', (_req, res) => res.status(200).send('ok'))
  const server = app.listen(0)
  const address = server.address()
  const port = typeof address === 'object' && address !== null ? address.port : 0
  return { server, url: `http://127.0.0.1:${port}/protected` }
}

async function get(url: string, headers: Record<string, string>): Promise<number> {
  const res = await fetch(url, { headers })
  return res.status
}

test('valid token is let through', async () => {
  const { server, url } = startTestApp()
  const token = jwt.sign({ sub: 'test-user' }, SECRET, { algorithm: 'HS256', expiresIn: '1h' })
  const status = await get(url, { Authorization: `Bearer ${token}` })
  server.close()
  assert.equal(status, 200)
})

test('missing Authorization header is rejected', async () => {
  const { server, url } = startTestApp()
  const status = await get(url, {})
  server.close()
  assert.equal(status, 401)
})

test('non-Bearer scheme is rejected', async () => {
  const { server, url } = startTestApp()
  const status = await get(url, { Authorization: 'Basic dXNlcjpwYXNz' })
  server.close()
  assert.equal(status, 401)
})

test('garbage token is rejected', async () => {
  const { server, url } = startTestApp()
  const status = await get(url, { Authorization: 'Bearer not.a.jwt' })
  server.close()
  assert.equal(status, 401)
})

test('token signed with the wrong secret is rejected', async () => {
  const { server, url } = startTestApp()
  const token = jwt.sign({ sub: 'test-user' }, 'wrong-secret', { algorithm: 'HS256', expiresIn: '1h' })
  const status = await get(url, { Authorization: `Bearer ${token}` })
  server.close()
  assert.equal(status, 401)
})

test('expired token is rejected', async () => {
  const { server, url } = startTestApp()
  const token = jwt.sign({ sub: 'test-user' }, SECRET, { algorithm: 'HS256', expiresIn: '-1h' })
  const status = await get(url, { Authorization: `Bearer ${token}` })
  server.close()
  assert.equal(status, 401)
})

test('alg:none token is rejected (algorithm confusion)', async () => {
  const { server, url } = startTestApp()
  const token = jwt.sign({ sub: 'test-user' }, null, { algorithm: 'none' })
  const status = await get(url, { Authorization: `Bearer ${token}` })
  server.close()
  assert.equal(status, 401)
})
